src/OptimizedPathMappingRepository.php

```diff
@@ -0,0 +1,365 @@
+ *
+ * (c) Bernhard Schussek <bschussek@gmail.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Puli\Repository;
+
+use Puli\Repository\Api\EditableRepository;
+use Puli\Repository\Api\Resource\FilesystemResource;
+use Puli\Repository\Api\ResourceCollection;
+use Puli\Repository\Resource\Collection\FilesystemResourceCollection;
+use Puli\Repository\Api\ResourceNotFoundException;
+use Puli\Repository\Api\UnsupportedLanguageException;
+use Puli\Repository\Api\UnsupportedResourceException;
+use Puli\Repository\Resource\DirectoryResource;
+use Puli\Repository\Resource\GenericFilesystemResource;
+use Puli\Repository\Resource\GenericResource;
+use Webmozart\Assert\Assert;
+use Webmozart\Glob\Iterator\GlobFilterIterator;
+use Webmozart\Glob\Iterator\RegexFilterIterator;
+use Webmozart\KeyValueStore\Api\KeyValueStore;
+use Webmozart\PathUtil\Path;
+
+/**
+ * An optimized path mapping resource repository.
+ * When a resource is added, all its children are resolved
+ * and getting them is much faster.
+ *
+ * Resources can be added with the method {@link add()}:
+ *
+ * ```php
+ * use Puli\Repository\OptimizedPathMappingRepository;
+ *
+ * $repo = new OptimizedPathMappingRepository();
+ * $repo->add('/css', new DirectoryResource('/path/to/project/res/css'));
+ * ```
+ *
+ * This repository only supports instances of FilesystemResource.
+ *
+ * @since  1.0
+ * @author Bernhard Schussek <bschussek@gmail.com>
+ * @author Titouan Galopin <galopintitouan@gmail.com>
+ */
+class OptimizedPathMappingRepository implements EditableRepository
+{
+    /**
+     * @var KeyValueStore
+     */
+    private $store;
+
+    /**
+     * Creates a new repository.
+     *
+     * @param KeyValueStore $store The store of all the paths.
+     */
+    public function __construct(KeyValueStore $store)
+    {
+        $this->store = $store;
+
+        $this->clear();
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function get($path)
+    {
+        Assert::stringNotEmpty($path, 'The path must be a non-empty string. Got: %s');
+        Assert::startsWith($path, '/', 'The path %s is not absolute.');
+
+        $path = Path::canonicalize($path);
+
+        if (! $this->store->exists($path)) {
+            throw ResourceNotFoundException::forPath($path);
+        }
+
+        return $this->store->get($path);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function find($query, $language = 'glob')
+    {
+        if ('glob' !== $language) {
+            throw UnsupportedLanguageException::forLanguage($language);
+        }
+
+        Assert::stringNotEmpty($query, 'The glob must be a non-empty string. Got: %s');
+        Assert::startsWith($query, '/', 'The glob %s is not absolute.');
+
+        $query = Path::canonicalize($query);
+        $resources = array();
+
+        if (false !== strpos($query, '*')) {
```

Comment:
You should use `Glob::isDynamic()` instead, as globs can contain other dynamic parts than `*` (e.g. `{foo,bar}`).